### PR TITLE
[FLINK-35363] FLIP-449: Reorganization of flink-connector-jdbc

### PR DESCRIPTION
**Pull request by eskabetxe**
This is first phase of FLIP
each task has its own commit

tasks done:
https://issues.apache.org/jira/browse/FLINK-35370
https://issues.apache.org/jira/browse/FLINK-35364
https://issues.apache.org/jira/browse/FLINK-35365
https://issues.apache.org/jira/browse/FLINK-35366
https://issues.apache.org/jira/browse/FLINK-35367
https://issues.apache.org/jira/browse/FLINK-35494

### CODE REVIEW

File: flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/table/JdbcFactory.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+ * Provider Interfaces (SPI) for discovering.
+ *
+ * <p>Classes that implement this interface can be added to the
+ * "META_INF/services/org.apache.flink.connector.jdbc.core.table.JdbcFactory" file of a JAR file in
+ * the current classpath to be found.
+ *
+ * @see JdbcDialect
+ */
+@PublicEvolving
+public interface JdbcFactory {
```
Comment (RocMarshal) [inline flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/table/JdbcFactory.java:37]: @eskabetxe Great work! 

Just left a few minor comments.
1 Maybe The original name is better for the followed items:
- Naming
- package structure
- The semantic of the class


2 The CI is failed , which is caused by something unknown possibly

Comment (eskabetxe) [reply flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/table/JdbcFactory.java:37]: point 1: I will change it as we talk offline..
point 2: Fixed, was a dependency problem//

Comment (eskabetxe) [reply flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/table/JdbcFactory.java]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/AbstractDialect.java [hunk 2]
```diff
@@ -41,228 +28,11 @@
  *
  * <p>Implementors should be careful to check the default SQL statements are performant for their
  * specific dialect and override them if necessary.
+ *
+ * @deprecated use org.apache.flink.connector.jdbc.core.table.dialect.AbstractDialect
```
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/AbstractDialect.java:32]:
```suggestion
 * @deprecated Use {@link org.apache.flink.connector.jdbc.core.database.dialect.AbstractDialect}.
```
How about using the 'link' tag to point the target new Class ?

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/AbstractJdbcCatalog.java [hunk 3]
```diff
@@ -18,551 +18,22 @@
...
-    private static final Logger LOG = LoggerFactory.getLogger(AbstractJdbcCatalog.class);
-
-    protected final ClassLoader userClassLoader;
-    protected final String baseUrl;
-    protected final String defaultUrl;
-    protected final Properties connectionProperties;
+/**
+ * Abstract catalog for any JDBC catalogs.
+ *
+ * @deprecated use org.apache.flink.connector.jdbc.core.table.catalog.AbstractJdbcCatalog
```
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/AbstractJdbcCatalog.java:24]:
```suggestion
 * @deprecated Use {@link org.apache.flink.connector.jdbc.core.database.catalog.AbstractJdbcCatalog}.
```

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/converter/AbstractJdbcRowConverter.java [hunk 4]
```diff
@@ -18,252 +18,17 @@
...
-    protected final JdbcDeserializationConverter[] toInternalConverters;
-    protected final JdbcSerializationConverter[] toExternalConverters;
-    protected final LogicalType[] fieldTypes;
-
-    public abstract String converterName();
 
+/**
+ * Base class for all converters that convert between JDBC object and Flink internal object.
+ *
+ * @deprecated use AbstractDialectConverter
```
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/converter/AbstractJdbcRowConverter.java:27]:
```suggestion
 * @deprecated Use {@link AbstractDialectConverter}.
```

File: flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/internal/options/InternalJdbcConnectionOptions.java [hunk 5]
```diff
@@ -179,7 +179,7 @@ public Builder setDBUrl(String dbURL) {
 
         /**
          * optional, Handle the SQL dialect of jdbc driver. If not set, it will be infer by {@link
-         * JdbcDialectLoader#load} from DB url.
+         * JdbcFactoryLoader#load} from DB url.
```
Comment (RocMarshal) [inline flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/internal/options/InternalJdbcConnectionOptions.java:182]:
```suggestion
         * JdbcFactoryLoader#load(String, ClassLoader)} from DB url.
```

File: flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/datastream/Jdbc.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+package org.apache.flink.connector.jdbc.core.datastream;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.connector.jdbc.core.datastream.sink.JdbcSink;
+import org.apache.flink.connector.jdbc.core.datastream.sink.JdbcSinkBuilder;
+import org.apache.flink.connector.jdbc.core.datastream.source.JdbcSource;
+import org.apache.flink.connector.jdbc.core.datastream.source.JdbcSourceBuilder;
+
+/** Facade to create JDBC stream sources and sinks. */
```
Comment (RocMarshal) [inline flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/datastream/Jdbc.java:26]: The repo is lacking of documentation pages to describe the new jdbc sink2 and the current usage for here.
Would you like to supplement it by https://issues.apache.org/jira/browse/FLINK-35811 ?

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/JdbcCatalog.java [hunk 7]
```diff
@@ -33,7 +34,12 @@
 
 import static org.apache.flink.connector.jdbc.JdbcConnectionOptions.getBriefAuthProperties;
 
-/** Catalogs for relational databases via JDBC. */
+/**
+ * Catalogs for relational databases via JDBC.
+ *
+ * @deprecated user org.apache.flink.connector.jdbc.core.table.catalog.JdbcCatalog
```
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/JdbcCatalog.java:40]:
```suggestion
 * @deprecated Use {@link org.apache.flink.connector.jdbc.core.database.catalog.JdbcCatalog}.
```

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/factory/JdbcCatalogFactory.java [hunk 8]
```diff
@@ -30,18 +31,26 @@
...
+import static org.apache.flink.connector.jdbc.core.database.catalog.factory.JdbcCatalogFactoryOptions.PASSWORD;
+import static org.apache.flink.connector.jdbc.core.database.catalog.factory.JdbcCatalogFactoryOptions.USERNAME;
 import static org.apache.flink.connector.jdbc.utils.JdbcUtils.getConnectionProperties;
 import static org.apache.flink.table.factories.FactoryUtil.PROPERTY_VERSION;
 
-/** Factory for {@link JdbcCatalog}. */
+/**
+ * Factory for {@link JdbcCatalog}.
+ *
+ * @deprecated use org.apache.flink.connector.jdbc.core.table.catalog.factory.JdbcCatalogFactory
```
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/factory/JdbcCatalogFactory.java:45]:
```suggestion
 * @deprecated Use {@link org.apache.flink.connector.jdbc.core.database.catalog.factory.JdbcCatalogFactory}.
```

File: flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/database/catalog/JdbcCatalogTypeMapper.java [hunk 9]
```diff
@@ -16,22 +16,19 @@
...
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
 
-    @Override
-    public JdbcDialect create() {
-        return new Db2Dialect();
-    }
+/** Separate the jdbc meta-information type to flink table type into the interface. */
+@PublicEvolving
+public interface JdbcCatalogTypeMapper {
```
Comment (RocMarshal) [inline flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/database/catalog/JdbcCatalogTypeMapper.java:30]: How about rename to 'JdbcTypeMapper' ?

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcConnectorOptions.java [hunk 10]
```diff
@@ -19,165 +19,13 @@
...
-import org.apache.flink.table.connector.source.lookup.LookupOptions;
-import org.apache.flink.table.factories.FactoryUtil;
 
-import java.time.Duration;
-
-/** Options for the JDBC connector. */
+/**
+ * Options for the JDBC connector.
+ *
+ * @deprecated please use {@link org.apache.flink.connector.jdbc.core.table.JdbcConnectorOptions}
```
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcConnectorOptions.java:26]:
```suggestion
 * @deprecated Please use {@link org.apache.flink.connector.jdbc.core.table.JdbcConnectorOptions}.
```

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java [hunk 11]
```diff
@@ -19,139 +19,16 @@
...
-import java.util.Optional;
+import org.apache.flink.connector.jdbc.core.database.JdbcFactory;
 
 /**
  * Represents a dialect of SQL implemented by a particular JDBC system. Dialects should be immutable
  * and stateless.
  *
- * @see JdbcDialectFactory
+ * @see JdbcFactory
+ * @deprecated use org.apache.flink.connector.jdbc.core.table.dialect.JdbcDialect
```
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:29]:
```suggestion
 * @deprecated Use {@link org.apache.flink.connector.jdbc.core.database.dialect.JdbcDialect}.
```

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialectFactory.java [hunk 12]
```diff
@@ -18,41 +18,34 @@
...
  *
  * <p>Classes that implement this interface can be added to the
- * "META_INF/services/org.apache.flink.connector.jdbc.dialect.JdbcDialectFactory" file of a JAR file
- * in the current classpath to be found.
+ * "META_INF/services/org.apache.flink.connector.jdbc.JdbcDialectFactory" file of a JAR file in the
+ * current classpath to be found.
  *
- * @see JdbcDialect
+ * @see org.apache.flink.connector.jdbc.dialect.JdbcDialect
+ * @deprecated use JdbcFactory
```
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialectFactory.java:33]:
```suggestion
 * @deprecated Use {@link JdbcFactory}.
```

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialectTypeMapper.java [hunk 13]
```diff
@@ -18,15 +18,12 @@
...
-/** Separate the jdbc meta-information type to flink table type into the interface. */
-public interface JdbcDialectTypeMapper {
-
-    DataType mapping(ObjectPath tablePath, ResultSetMetaData metadata, int colIndex)
-            throws SQLException;
-}
+/**
+ * Separate the jdbc meta-information type to flink table type into the interface.
+ *
+ * @deprecated use JdbcCatalogTypeMapper
```
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialectTypeMapper.java:26]:
```suggestion
 * @deprecated Use {@link JdbcCatalogTypeMapper}.
```

Maybe it need the renamed name mentioned above~

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/converter/JdbcRowConverter.java [hunk 14]
```diff
@@ -19,34 +19,15 @@
...
 
-import java.io.Serializable;
-import java.sql.ResultSet;
-import java.sql.SQLException;
-
 /**
  * Converter that is responsible to convert between JDBC object and Flink SQL internal data
  * structure {@link RowData}.
+ *
+ * @deprecated use JdbcDialectConverter
```
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/converter/JdbcRowConverter.java:29]:
```suggestion
 * @deprecated Use {@link JdbcDialectConverter}.
```

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/JdbcSource.java [hunk 15]
```diff
@@ -50,8 +50,13 @@
 import java.util.ArrayList;
 import java.util.Objects;
 
-/** JDBC source. */
+/**
+ * JDBC source.
+ *
+ * @deprecated please use {@link org.apache.flink.connector.jdbc.core.datastream.source.JdbcSource}
```
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/JdbcSource.java:56]:
```suggestion
 * @deprecated Please use {@link org.apache.flink.connector.jdbc.core.datastream.source.JdbcSource}.
```

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/JdbcSourceBuilder.java [hunk 16]
```diff
@@ -94,8 +89,11 @@
  * @see PreparedStatement
  * @see DriverManager
  * @see JdbcSource
+ * @deprecated please use {@link
+ *     org.apache.flink.connector.jdbc.core.datastream.source.JdbcSourceBuilder}
```
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/JdbcSourceBuilder.java:93]:
```suggestion
 * @deprecated Please use {@link
 *     org.apache.flink.connector.jdbc.core.datastream.source.JdbcSourceBuilder}.
```

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/reader/extractor/ResultExtractor.java [hunk 17]
```diff
@@ -19,39 +19,19 @@
...
 
-import java.io.Serializable;
 import java.sql.ResultSet;
-import java.sql.SQLException;
 
 /**
  * The Extractor to extract the data from {@link ResultSet}.
  *
  * @param <T> The target data type.
+ * @deprecated please use {@link
```
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/reader/extractor/ResultExtractor.java:29]:
```suggestion
 * @deprecated Please use {@link
```

**Review by RocMarshal — COMMENTED** [review]
Thanks @eskabetxe for the hard work!
Looks nice to me on the functional refactor.
Just left a few of comments about doc/naming.
Please let me know what's your opinion :)

File: flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/datastream/Jdbc.java (hunk 6 above)
Comment (eskabetxe) [reply flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/datastream/Jdbc.java:26]: I think that all doc should also be refactored, maybe have one sub-page per connector, and clean all deprecated docs.
We could start talk about it, but it is on the plan.

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/AbstractDialect.java (hunk 2 above)
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/AbstractDialect.java:32]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/converter/AbstractJdbcRowConverter.java (hunk 4 above)
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/converter/AbstractJdbcRowConverter.java:27]: done

File: flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/internal/options/InternalJdbcConnectionOptions.java (hunk 5 above)
Comment (eskabetxe) [reply flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/internal/options/InternalJdbcConnectionOptions.java:182]: changed to JdbcFactoryLoader#loadDialect

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/JdbcCatalog.java (hunk 7 above)
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/JdbcCatalog.java:40]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/factory/JdbcCatalogFactory.java (hunk 8 above)
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/factory/JdbcCatalogFactory.java:45]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcConnectorOptions.java (hunk 10 above)
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcConnectorOptions.java:26]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java (hunk 11 above)
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:29]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialectFactory.java (hunk 12 above)
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialectFactory.java:33]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialectTypeMapper.java (hunk 13 above)
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialectTypeMapper.java:26]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/converter/JdbcRowConverter.java (hunk 14 above)
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/converter/JdbcRowConverter.java:29]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/JdbcSource.java (hunk 15 above)
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/JdbcSource.java:56]: fixed

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/JdbcSourceBuilder.java (hunk 16 above)
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/JdbcSourceBuilder.java:93]: fixes

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/reader/extractor/ResultExtractor.java (hunk 17 above)
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/reader/extractor/ResultExtractor.java:29]: fixed

File: flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/database/catalog/JdbcCatalogTypeMapper.java (hunk 9 above)
Comment (eskabetxe) [reply flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/database/catalog/JdbcCatalogTypeMapper.java:30]: we use the same pattern on all Jdbc{kink}Anything, where kind is source, sink, dialect, catalog (already from before the refactor)

this was called "JdbcDialectTypeMapper"

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/AbstractJdbcCatalog.java (hunk 3 above)
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/AbstractJdbcCatalog.java]: done

**Comment by eskabetxe** [comment]
@leonardBang @1996fanrui @snuyanzin can you check please when you have time

70% of the refactor has been completed.
The deprecation of the old sources and sinks is still pending, as they are still used in the table API. 
The idea is to review and merge this part (there are already significant changes) and continue working on other PRs to:

- Modify the table API for the new sources and sinks.
- Deprecate and move the old sources and sinks.
- Finish organizing the code in the core module.
- Reorganize the documentation.

Note: Before the merge, we can consider removing the backward-compatibility module to avoid generating the jar, as it was only meant to be a helper.

**Comment by RocMarshal** [comment]
> Note: Before the merge, we can consider removing the backward-compatibility module (and workflow) to avoid generating the jar, as it was only meant to be a helper.

Hi, @eskabetxe 
Considering that this is a relatively large refactoring action,
Perhaps retaining this module can also help developers quickly verify and fix compatibility related issues.
How about continuing to keep one or two versions of the cycle? In other words, keep this module until the refactored code becomes stable and mature.
Please let me know what's your opinion.

**Review by RocMarshal — APPROVED** [review]
Thanks @eskabetxe for the hard work.
LGTM +1. 
now ping @1996fanrui

**Comment by 1996fanrui** [comment]
Thanks @eskabetxe for the contribution, and thanks @RocMarshal for the review!

Due to both of you are active contributor of [flink-connector-jdbc](https://github.com/apache/flink-connector-jdbc), and you aren't committer. So I will merge this PR if no objection before next Monday(in 5 days).

**Pull request merged**